Enable oci hooks through API settings

arnaldo2792 opened the pull request:
**Issue number:**
N / A

**Description of changes:**

```
api: add oci-hooks setting to enable OCI hooks
```

The oci-hooks setting allows a user to enable OCI hooks provided by the OS. For the time being, the only OCI hook provided is the `log4j2-hotpatch`, which applies the hotpatch for Apache Log4j2 to containers running JVMs.

Remaining work:
- [x] Documentation for the new setting
- [x] Add API setting to vmware variants
- [x] Add migrations
- [x] Upgrade/Downgrade migration test

**Testing done:**
In aws-ecs/aws-k8s-1.21:

Run workloads with java services, with and without enabling the `log4j-hotpatch` hook. I saw the containers being patched, and the logs generated in `/dev/shm/hotdog.log`.

- [x] Run testing pods in aws k8s variants
- [ ] Run testing pods in vmware k8s variants

**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

samuelkarp reviewed (commented):
This provisionally looks good to me.

cbgbt commented:
We need a migration to remove this setting on downgrade, as discussed out of github. But this otherwise LGTM.

webern commented on packages/os/os.spec:
```diff
@@ -1,7 +1,6 @@
 %global _cross_first_party 1
 %global _is_k8s_variant %(if echo %{_cross_variant} | grep -Fqw "k8s"; then echo 1; else echo 0; fi)
 %global _is_aws_variant %(if echo %{_cross_variant} | grep -Fqw "aws"; then echo 1; else echo 0; fi)
-%global _is_vendor_variant %(if echo %{_cross_variant} | grep -Fqw "nvidia"; then echo 1; else echo 0; fi)
```
Is this on develop?

arnaldo2792 replied:
It was, yeah, the line was added with shimpei, because we didn't want to use it as the default runtime for all variants. But now we do :+1:

webern reviewed (commented):
Don't we need a migration?

webern commented:
> We need a migration to remove this setting on downgrade, as discussed out of github. But this otherwise LGTM.

Oh, things are happening to fast, I can't keep up 😅

arnaldo2792 commented:
Forced push includes:

- Missing settings for vmware variants
- Removed restart services configurations

arnaldo2792 commented:
Forced push includes rebase to upstream.

arnaldo2792 committed:
packages: set shimpei as default runtime for k8s and docker

With shimpei as the default runtime for container runtimes, we can run
OCI hooks provided through API settings. This commit needs the OCI
settings to properly work, since shimpei reads a hooks file generated by
changes in the settings.

Signed-off-by: Arnaldo Garcia Rincon <agarrcia@amazon.com>

arnaldo2792 committed:
api: add oci-hooks setting to enable OCI hooks

The oci-hooks setting allows a user to enable OCI hooks provided by the
OS. For the time being, the only OCI hook provided is the
`log4j2-hotpatch`, which applies the hotpatch for Apache Log4j2 to
containers running JVMs.

Signed-off-by: Arnaldo Garcia Rincon <agarrcia@amazon.com>

arnaldo2792 committed:
packages: add log4j-hotpatch to release

We always include log4j-hotpatch in all variants

Signed-off-by: Arnaldo Garcia Rincon <agarrcia@amazon.com>

arnaldo2792 commented:
Forced push includes:

- Simplified configuration files, based on offline conversation with @bcressey 
- Migrations for the new API
- Documentation for the new API
- Missing dependency in `release` for `log4j-hotpatch`

bcressey commented on README.md:
```diff
@@ -418,6 +418,13 @@ These settings can be changed at any time.
+#### OCI Hooks settings
+
+Bottlerocket allows you to opt-in to use additional [OCI hooks](https://github.com/opencontainers/runtime-spec/blob/main/runtime.md#lifecycle) for your orchestrated containers.
+Once you opt-in to use additional OCI hooks, the new orchestrated containers will be configured with them, but the existing containers won't be changed.
```
```suggestion
Once you opt-in to use additional OCI hooks, any new orchestrated containers will be configured with them, but existing containers won't be changed.nit:
```

cbgbt replied:
nit:
```suggestion
Once you opt-in to using additional OCI hooks, newly-orchestrated containers will be configured with them. Existing containers won't be changed.
```

bcressey commented on packages/release/Cargo.toml:
```diff
@@ -35,12 +35,9 @@ kexec-tools = { path = "../../packages/kexec-tools" }
 libaudit = { path = "../libaudit" }
 libgcc = { path = "../libgcc" }
 libstd-rust = { path = "../libstd-rust" }
+log4j2-hotpatch= { path = "../log4j2-hotpatch" }
```
nit: this seems like more properly a dependency of `hotdog` which is what consumes the hot patch jar

bcressey commented on packages/release/release.spec:
```diff
@@ -82,6 +82,7 @@ Requires: %{_cross_os}hotdog
 Requires: %{_cross_os}iproute
 Requires: %{_cross_os}iptables
 Requires: %{_cross_os}kexec-tools
+Requires: %{_cross_os}log4j2-hotpatch
```
nit: likewise, maybe better as a `hotdog` package dependency

bcressey commented on sources/models/shared-defaults/defaults.toml:
```diff
@@ -149,3 +149,11 @@ restart-commands = ["/usr/bin/certdog"]
 [metadata.settings.pki]
 affected-services = ["pki"]
+
+# OCI hooks
+
+[settings.oci-hooks]
+log4j-hotpatch-enabled = false
+
+[metadata.settings.oci-hooks]
+affected-services = ["oci-hooks"]
```
I'd probably put all this in `shared-defaults/oci-hooks.toml`, because you're going to want a different set of affected services for `shared-defaults/nvidia-oci-hooks.toml` when that shows up.

cbgbt reviewed (commented):
LGTM, pending testing! 🚢

arnaldo2792 commented:
Forced push includes:

- Fixed documentation
- Moved default settings to `oci-hooks.toml`
- Add `log4j-hotpatch` as a dependency of `hotdog`
- Changed commit's order

cbgbt reviewed (approved):
:shipit:
